storage/storage_image.go

```diff
@@ -93,309 +90,655 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
+	if err != nil && errors.Cause(err) != storage.ErrLayerUnknown {
+		return false, -1, errors.Wrapf(err, `Error looking for layers with digest %q`, blobinfo.Digest)
+	}
+	if len(layers) > 0 {
+		// Save this for completeness.
+		s.blobDiffIDs[blobinfo.Digest] = layers[0].UncompressedDigest
+		return true, layers[0].UncompressedSize, nil
+	}
+	// Check if we have a was-compressed layer in storage that's based on that blob.
+	layers, err = s.imageRef.transport.store.LayersByCompressedDigest(blobinfo.Digest)
+	if err != nil && errors.Cause(err) != storage.ErrLayerUnknown {
+		return false, -1, errors.Wrapf(err, `Error looking for compressed layers with digest %q`, blobinfo.Digest)
+	}
+	if len(layers) > 0 {
+		// Record the uncompressed value so that we can use it to calculate layer IDs.
+		s.blobDiffIDs[blobinfo.Digest] = layers[0].UncompressedDigest
+		return true, layers[0].CompressedSize, nil
+	}
+	// Nope, we don't have it.
 	return false, -1, nil
 }
 
+// ReapplyBlob is now a no-op, assuming HasBlob() says we already have it, since Commit() can just apply the
+// same one when it walks the list in the manifest.
 func (s *storageImageDestination) ReapplyBlob(blobinfo types.BlobInfo) (types.BlobInfo, error) {
-	err := blobinfo.Digest.Validate()
+	present, size, err := s.HasBlob(blobinfo)
+	if !present {
+		return types.BlobInfo{}, errors.Errorf("error reapplying blob %+v: blob was not previously applied", blobinfo)
+	}
 	if err != nil {
-		return types.BlobInfo{}, err
+		return types.BlobInfo{}, errors.Wrapf(err, "error reapplying blob %+v", blobinfo)
 	}
-	if layerList, ok := s.Layers[blobinfo.Digest]; !ok || len(layerList) < 1 {
-		b, err := s.imageRef.transport.store.ImageBigData(s.ID, blobinfo.Digest.String())
+	blobinfo.Size = size
+	return blobinfo, nil
+}
+
+// computeID computes a recommended image ID based on information we have so far.
+func (s *storageImageDestination) computeID(m manifest.Manifest) string {
+	mb, err := m.Serialize()
+	if err != nil {
+		return ""
+	}
+	switch manifest.GuessMIMEType(mb) {
+	case manifest.DockerV2Schema2MediaType, imgspecv1.MediaTypeImageManifest:
+		// For Schema2 and OCI1(?), the ID is just the hex part of the digest of the config blob.
+		logrus.Debugf("trivial image ID for configured blob")
+		configInfo := m.ConfigInfo()
+		if configInfo.Digest.Validate() == nil {
+			return configInfo.Digest.Hex()
+		}
+		return ""
+	case manifest.DockerV2Schema1MediaType, manifest.DockerV2Schema1SignedMediaType:
+		// Convert the schema 1 compat info into a schema 2 config, constructing some of the fields
+		// that aren't directly comparable using info from the manifest.
+		logrus.Debugf("computing image ID using compat data")
+		s1, ok := m.(*manifest.Schema1)
+		if !ok {
+			logrus.Debugf("schema type was guessed wrong?")
+			return ""
+		}
+		if len(s1.History) == 0 {
+			logrus.Debugf("image has no layers")
+			return ""
+		}
+		s2 := struct {
```

Comment:
That type is what the `github.com/docker/distribution/manifest/schema1/config_builder.go` logic builds.  This definition keeps us from discarding parts of the `Config` structure, which can include more details than that logic knows how to encode (perhaps having been generated by a toolchain that doesn't exist any more?), so that we can incorporate its contents into the ID calculation.